### PR TITLE
Add c++ gpu kernel wrappers & simplify wave_functions

### PR DESCRIPTION
**Pull request by haampie**
- Add kernel wrappers, dry code, make runtime error compile time failure
- ignore more spack build folders

### CODE REVIEW

**Comment by haampie** [comment]
bors try

File: src/SDDK/wave_functions.cpp [hunk 1]
```diff
@@ -411,31 +416,15 @@ mdarray<T, 1> Wave_functions<T>::sumsqr(device_t pu__, spin_range spins__, int n
...
             case device_t::GPU: {
 #if defined(SIRIUS_GPU)
-                if(std::is_same<T, double>::value) {
-                    add_square_sum_gpu_double(
-                        reinterpret_cast<std::complex<double> const*>(pw_coeffs(is).prime().at(memory_t::device)),
-                        pw_coeffs(is).num_rows_loc(), n__, gkvecp_.gvec().reduced(), comm_.rank(),
+                add_square_sum_gpu(
+                    reinterpret_cast<std::complex<T> const*>(pw_coeffs(is).prime().at(memory_t::device)),
+                    pw_coeffs(is).num_rows_loc(), n__, gkvecp_.gvec().reduced(), comm_.rank(),
+                    reinterpret_cast<double*>(s.at(memory_t::device)));
```
Comment (AdhocMan) [inline src/SDDK/wave_functions.cpp:422]:
```suggestion
                    reinterpret_cast<T*>(s.at(memory_t::device)));
```
The same at the next call.
Maybe casting is not even required, depending on what types `pw_coeffs` and `s` hold.

**Review by AdhocMan — CHANGES_REQUESTED** [review]

File: src/SDDK/wave_functions.cpp (hunk 1 above)
Comment (haampie) [reply src/SDDK/wave_functions.cpp]: Yeah, that's much better.

**Review by AdhocMan — APPROVED** [review]

**Review by manhin321 — APPROVED** [review]

**Comment by toxa81** [comment]
bors try

**Comment by haampie** [comment]
jfrog is down...

**Comment by haampie** [comment]
bors try

**Pull request merged**